### PR TITLE
Progress Indicator Overhaul

### PR DESCRIPTION
**Pull request by szabolcs-instructure**
Rewritten and unified progress indicators across the whole application. 

**SwiftUI indicators** 
- Built in `ProgressView` indicator with the following available styles 
  - DeterminateBarProgressStyle
  - IndeterminateBarProgressStyle
  - DeterminateCircleProgressStyle
  - IndeterminateCircleProgressStyle

- `RefreshableScrollView` 

---

**UIKit indicators**
- `CircleRefreshControl` which uses `CircleProgressView` under the hood

---

refs: MBL-16238
affects: Student, Teacher
release note: New loading indicator for lists and webViews
test plan: See ticket

## Screenshots
<table>
<tr><th>Dashboard</th><th>K5Dashboard</th></tr>
<tr>
<td></td>
<td></td>
</tr>
<tr><th>Pages</th><th>K5Modules</th></tr>
<tr>
<td></td>
<td></td>
</tr>
</table>

## Checklist

- [x] Follow-up e2e test ticket created or not needed
- [x] A11y checked
- [x] Tested on phone
- [x] Tested on tablet
- [x] Approve from product or not needed

### CODE REVIEW

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Build failed, skipping coverage check
  </td>
    </tr>
  </tbody>
</table>



<details><summary>:x: Swift lint</summary>


    ❌ /Users/vagrant/git/Core/Core/Dashboard/K5/View/ImportantDates/K5ImportantDatesView.swift:95:80: Trailing Comma Violation: Multi-line collection literals should have trailing commas. (trailing_comma)
    /Users/vagrant/git/Core/Core/Dashboard/K5/View/ImportantDates/K5ImportantDatesView.swift:102:75: error: Trailing Comma Violation: Multi-line collection literals should have trailing commas. (trailing_comma)
    
</details>


[MBL-16238](https://instructure.atlassian.net/browse/MBL-16238)

#### Affected Apps: Student, Teacher

#### Release Note: 
New loading indicator for lists and webViews
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by inst-danger** [comment]
<details><summary>Teacher Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<details><summary>Student Build QR Code:</summary>

</details>

File: Core/Core/SwiftUIViews/RefreshableScrollView.swift [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+            content: {
+                VStack(spacing: 0) {
+                    switch viewState {
+                    case .animating:
+                        indeterminateProgressView()
+                    case let .progress(progress):
+                        determinateProgressView(progress)
+                    }
+                    content()
+                        .offset(x: 0, y: isAnimating ? 0 : -48)
```
Comment (vargaat) [inline Core/Core/SwiftUIViews/RefreshableScrollView.swift:70]: We could extract these constants to be meaningful. I guess this 48 is the indeterminateCircle factory method's default 32 point size plus the 16 point padding above the progress circle.

File: Core/Core/Dashboard/K5/View/ImportantDates/K5ImportantDatesView.swift [hunk 2]
```diff
@@ -71,39 +77,42 @@ public struct K5ImportantDatesView: View {
...
-                            events: [K5ImportantDateItem(subject: "History",
-                                                         title: "This important history event",
-                                                         color: .green,
-                                                         date: Date(fromISOString: "2022-01-03T08:00:00Z")!,
-                                                         route: nil,
-                                                         type: .event), ]),
-        ]
-        return K5ImportantDatesViewModel(with: dates)
-    }
+    struct K5ImportantDates_Previews: PreviewProvider {
```
Comment (vargaat) [inline Core/Core/Dashboard/K5/View/ImportantDates/K5ImportantDatesView.swift:80]: Why the indentation?

**Review by vargaat — CHANGES_REQUESTED** [review]
Good job in overall!

- [x] I think we should add a bottom padding to the circle because if we refresh content without any top padding than it'll look strange. <br/>
- [ ] I wouldn't call this a bug but I think it's strange: If I keep the content pulled down and move it up and down around the snapping point then it will keep refreshing as long as I do so. Maybe we should make the next refresh only trigger if the content offset reached 0 after the refresh. 
https://user-images.githubusercontent.com/72396990/193078475-4d4659fa-74f2-444f-bfbb-d1f7174e5c55.MP4

File: Core/Core/SwiftUIViews/RefreshableScrollView.swift [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    // MARK: - Init
+
+    public init(
+        _ axes: Axis.Set = .vertical,
+        showsIndicators: Bool = true,
+        @ViewBuilder content: @escaping () -> Content,
+        refreshAction: @escaping (@escaping () -> Void) -> Void
+    ) {
+        self.axes = axes
+        showsIndicator = showsIndicators
```
[the remark this answers is not in the document]
Comment (szabolcs-instructure) [reply Core/Core/SwiftUIViews/RefreshableScrollView.swift:52]: Typo, fixed!

File: Core/Core/SwiftUIViews/RefreshableScrollView.swift [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+import SwiftUI
+
+public struct RefreshableScrollView<Content: View>: View {
+    enum ViewState {
```
[the remark this answers is not in the document]
Comment (szabolcs-instructure) [reply Core/Core/SwiftUIViews/RefreshableScrollView.swift:22]: Changed

File: Core/Core/SwiftUIViews/ProgressViewStyle/DeterminateBarProgressViewStyle.swift [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    // MARK: - Dependencies
+
+    private let foregroundColor: Color
+    private let backgroundColor: Color
+
+    // MARK: - Private properties
+
+    private let barHeight: CGFloat = 4
+    private let barWidthPercentage: CGFloat = 0.4
+    private let maxOffset: Float = 1 / 0.4
```
[the remark this answers is not in the document]
Comment (szabolcs-instructure) [reply Core/Core/SwiftUIViews/ProgressViewStyle/DeterminateBarProgressViewStyle.swift:31]: Removed

File: Core/Core/Dashboard/K5/View/ImportantDates/K5ImportantDatesView.swift (hunk 2 above)
Comment (szabolcs-instructure) [reply Core/Core/Dashboard/K5/View/ImportantDates/K5ImportantDatesView.swift:80]: SwiftLint was giving a warning

File: Core/Core/SwiftUIViews/RefreshableScrollView.swift (hunk 1 above)
Comment (szabolcs-instructure) [reply Core/Core/SwiftUIViews/RefreshableScrollView.swift:70]: Changed!

File: Canvas.xcworkspace/contents.xcworkspacedata [hunk 6]
```diff
@@ -80,9 +80,6 @@
       <FileRef
          location = "group:NightlyTests.xctestplan">
       </FileRef>
-      <FileRef
```
[the remark this answers is not in the document]
Comment (szabolcs-instructure) [reply Canvas.xcworkspace/contents.xcworkspacedata]: No idea how this change occurred :D. I reverted it.

File: Core/Core/Files/FileUploadProgress/View/FileProgressListView.swift [hunk 7]
```diff
@@ -142,12 +142,24 @@ struct FileProgressListView<ViewModel>: View where ViewModel: FileProgressListVi
...
-            progress: value,
-            foregroundColor: Color(Brand.shared.primary),
-            backgroundColor: Color(Brand.shared.primary).opacity(0.2)
-        )
+    @ViewBuilder private func progressView(value: Float? = nil) -> some View {
+        if let progress = value {
+            ProgressView(value: progress)
+                .progressViewStyle(
+                    .determinateBar(
+                        foregroundColor: Color(Brand.shared.primary),
```
[the remark this answers is not in the document]
Comment (szabolcs-instructure) [reply Core/Core/Files/FileUploadProgress/View/FileProgressListView.swift:150]: Fixed

File: Core/Core/Dashboard/K5/View/Resources/K5ResourcesView.swift [hunk 8]
```diff
@@ -27,11 +27,7 @@ public struct K5ResourcesView: View {
     }
 
     public var body: some View {
-        ScrollView(.vertical, showsIndicators: false) {
-            CircleRefresh { endRefreshing in
-                viewModel.refresh(completion: endRefreshing)
-            }
-
+        RefreshableScrollView {
```
[the remark this answers is not in the document]
Comment (szabolcs-instructure) [reply Core/Core/Dashboard/K5/View/Resources/K5ResourcesView.swift:30]: Fixed!

**Comment by szabolcs-instructure** [comment]
> Content on the Homeroom page is moving up and down a bit during refreshing, I wonder if that's avoidable.

I double checked and noticed that this behaviour is only present on iOS 14. Since we are planning to bump the minimum version to iOS 15 soon, there won't be any fixes for this issue.

**Review by vargaat — APPROVED** [review]
QA+1

**Review by idevainstr — APPROVED** [review]
QA +1

**Pull request merged**